File: Packages/TypeEngine/Sources/TypeEngine/TypeEngine.swift
import Foundation
import LemmaCore
import Lexicon

extension BinaryLemmatizer: MorphologyProviding {}

/// Facade over the corrector and predictor with a running bilingual language
/// posterior. Pure synchronous API — no dispatch/async inside; the caller
/// owns threading. Not thread-safe (confirmWord mutates the posterior); use
/// from a single queue.
public final class TypeEngine {
    public let config: EngineConfig

    private let model: BlendedLanguageModel
    private let corrector: Corrector
    private let predictor: Predictor

    /// Lane posterior P(lane = Icelandic): the forward probability of the
    /// two-state IS/EN switching model over committed words, clamped to
    /// [posteriorFloor, posteriorCeiling] so it never saturates. Starts at
    /// the neutral 0.5 prior.
    public private(set) var probabilityIcelandic: Double

    /// Production initializer: BÍN morphology via LemmaCore.
    public convenience init(
        icelandic: Lexicon,
        english: Lexicon,
        morphology: BinaryLemmatizer?,
        config: EngineConfig = EngineConfig()
    ) {
        self.init(
            icelandic: icelandic,
            english: english,
            morphologyProvider: morphology,
            config: config
        )
    }

    /// Designated initializer; accepts any MorphologyProviding (tests use a
    /// dictionary-backed fake).
    public init(
        icelandic: Lexicon,
        english: Lexicon,
        morphologyProvider: MorphologyProviding? = nil,
        config: EngineConfig = EngineConfig()
    ) {
        self.config = config
        let model = BlendedLanguageModel(
            icelandic: icelandic,
            english: english,
            morphology: morphologyProvider,
            config: config,
            personal: PersonalStore()
        )
        self.model = model
        self.corrector = Corrector(model: model, config: config)
        self.predictor = Predictor(model: model, config: config)
        self.probabilityIcelandic = 0.5
    }

    // MARK: - Personal vocabulary (M2 learning)

    /// Inject (or clear) the personal-learning snapshot. Cheap O(personal
    /// scale) index rebuild — no recalibration, no engine rebuild — so the
    /// embedder can swap freely whenever the model file changes on disk.
    /// Must be called on the queue that owns this engine (same confinement
    /// contract as every other engine call); the in-session learned overlay
    /// survives the swap (a fresh snapshot may or may not already contain
    /// those words — double counting is bounded and harmless for a boost).
    public func setPersonalVocabulary(_ vocabulary: PersonalVocabulary?) {
        model.personal.setSnapshot(vocabulary)
    }

    /// Session-immediate learning: make `word` valid + suggestible RIGHT NOW
    /// (verbatim tap / explicit learn signals must not wait for the app's
    /// compaction). The overlay lives until `clearSessionVocabulary()`.
    public func learnSessionWord(_ word: String) {
        model.personal.learnSession(word)
    }

    /// Drop the in-session learned overlay (session reset).
    public func clearSessionVocabulary() {
        model.personal.clearSession()
    }

    /// Diagnostics (REPL `:learned`, tests): canonical surfaces of the
    /// injected snapshot's words, sorted.
    public var personalSnapshotWords: [String] { model.personal.snapshotWords }

    /// Diagnostics: words learned in this session (overlay), sorted.
    public var sessionLearnedWords: [String] { model.personal.sessionWords }

    /// Whether `word` is currently valid personal vocabulary (snapshot or
    /// session overlay), case-insensitively.
    public func isPersonalWord(_ word: String) -> Bool {
        model.personal.isValidWord(word)
    }

    /// Suggestions for the suggestion bar.
    ///
    /// - Parameters:
    ///   - context: committed text before the current word (used only for the
    ///     trailing word, as bigram context)
    ///   - currentWord: the word being typed; empty string means "predict the
    ///     next word"
    ///   - limit: maximum suggestions returned
    ///   - deliberateCharacters: characters of the current word that were
    ///     entered via long-press/callout (the strongest deliberateness
    ///     signal — vetoes lane-relaxation folding for this word; see
    ///     `Corrector.correct` and `TypingSession.noteLongPressInsertion`)
    ///   - taps: per-position touch samples aligned with `currentWord`
    ///     (PLAN.md "Touch decoding", stage 1; see `Corrector.correct`).
    ///     Empty = static spatial pricing, unchanged.
    public func suggestions(
        context: String,
        currentWord: String,
        limit: Int = 3,
        deliberateCharacters: [Character] = [],
        taps: [TapSample?] = []
    ) -> [Suggestion] {
        let previous = Self.lastWord(in: context)
        let trimmed = currentWord.trimmingCharacters(in: .whitespacesAndNewlines)

        if trimmed.isEmpty {
            return restorePersonalSurfaces(
                predictor.nextWords(
                    previousWord: previous,
                    pIcelandic: probabilityIcelandic,
                    limit: limit
                )
            )
        }

        let result = corrector.correct(
            typed: trimmed,
            previousWord: previous,
            pIcelandic: probabilityIcelandic,
            limit: limit,
            deliberateCharacters: deliberateCharacters,
            taps: taps.count == trimmed.count ? taps : []
        )
        var suggestions = restorePersonalSurfaces(result.suggestions)

        // Preserve the user's leading capitalization.
        if let first = trimmed.first, first.isUppercase {
            suggestions = suggestions.map {
                Suggestion(
                    text: $0.text.prefix(1).uppercased() + $0.text.dropFirst(),
                    isAutocorrect: $0.isAutocorrect,
                    confidence: $0.confidence,
                    isRestoration: $0.isRestoration
                )
            }
        }
        // An "autocorrect" into the byte-identical typed text is a no-op
        // and must not be flagged (surfaces via the case-insensitive
        // pipeline: typed "I" → the lone-i capitalization suggestion "I").
        return suggestions.map {
            $0.isAutocorrect && $0.text == trimmed
                ? Suggestion(
                    text: $0.text, isAutocorrect: false, confidence: $0.confidence,
                    isRestoration: $0.isRestoration)
                : $0
        }
    }

    /// Dotted-token space-miss escape (PLAN.md "Space-miss correction" +
    /// verbatim/URL layers; dogfood "sem.er"): the "left right" reading of
    /// a word.word token whose SHAPE already passed TypingSession's URL
    /// checks. Returns nil unless both halves are common attested words in
    /// one common language; `isAutocorrect` follows the strict split rules
    /// (see `Corrector.dotSplitSuggestion`).
    public func dottedSpaceMiss(left: String, right: String, context: String) -> Suggestion? {
        let previous = Self.lastWord(in: context)
        guard
            let suggestion = corrector.dotSplitSuggestion(
                left: left.lowercased(),
                right: right.lowercased(),
                previousWord: previous,
                pIcelandic: probabilityIcelandic
            )
        else { return nil }
        // Preserve the user's leading capitalization ("Sem.er" → "Sem er").
        if let first = left.first, first.isUppercase {
            return Suggestion(
                text: suggestion.text.prefix(1).uppercased() + suggestion.text.dropFirst(),
                isAutocorrect: suggestion.isAutocorrect,
                confidence: suggestion.confidence
            )
        }
        return suggestion
    }

    /// Personal surface forms are byte-exact and case-preserving
    /// ("Miðeind") while the correction/prediction pipeline works in
    /// lowercase — restore the canonical personal casing on the way out.
    private func restorePersonalSurfaces(_ suggestions: [Suggestion]) -> [Suggestion] {
        suggestions.map { suggestion in
            guard let surface = model.personal.displaySurface(of: suggestion.text) else {
                return suggestion
            }
            return Suggestion(
                text: surface,
                isAutocorrect: suggestion.isAutocorrect,
                confidence: suggestion.confidence,
                isRestoration: suggestion.isRestoration
            )
        }
    }

    /// Update the language posterior after the user commits a word
    /// (typed through, tapped a suggestion, or accepted an autocorrect).
    ///
    /// One forward step of the two-lane (IS/EN) switching model — typing is
    /// either Icelandic-with-slettur or English, lanes are sticky but not
    /// strict (PLAN.md "Bilingual blending — lane model"):
    ///
    ///   predict: p' = (1-s)·p + s·(1-p)          s = laneSwitchProbability
    ///   update:  p  = p'·e_IS / (p'·e_IS + (1-p')·e_EN)
    ///
    /// where log(e_IS/e_EN) is the word's graded emission evidence from the
    /// calibrated per-lexicon z-scores (see BlendedLanguageModel.laneEvidence).
    /// The capped evidence + low switch prior make one sletta a bounded
    /// nudge (the lane holds) while 2–3 consecutive off-lane words flip it.
    /// Words with no attributable evidence (OOV, junk-tier like "dont" in
    /// is.lex web noise, ambiguous both-lexicon words) have ~uniform
    /// emissions: they only apply the predict step, i.e. a gentle decay of
    /// the lane toward neutral — never a drag toward either language.
    public func confirmWord(_ word: String) {
        let w = word.lowercased().trimmingCharacters(in: .whitespacesAndNewlines)
        guard !w.isEmpty else { return }

        let s = config.laneSwitchProbability
        let predicted = (1 - s) * probabilityIcelandic + s * (1 - probabilityIcelandic)
        let evidence = model.laneEvidence(of: w)
        let odds = (predicted / (1 - predicted)) * exp(evidence)
        let updated = odds / (1 + odds)
        probabilityIcelandic = min(max(updated, config.posteriorFloor), config.posteriorCeiling)
    }

    /// Relax the lane posterior toward the neutral 0.5 prior across a
    /// sentence boundary (". ", "!", "?"): sheds laneBoundaryDecay of the
    /// distance to neutral — the lane weakens with discourse distance but
    /// does not reset. TypingSession calls this when a committed word's
    /// trailing delimiter is a sentence terminator.
    public func noteSentenceBoundary() {
        let decay = config.laneBoundaryDecay
        probabilityIcelandic = 0.5 + (probabilityIcelandic - 0.5) * (1 - decay)
    }

    /// Lane-model diagnostics for a word (REPL `:word` command): per-lexicon
    /// attestation + calibrated z-scores, BÍN morphology validity, and the
    /// resulting emission evidence log(e_IS/e_EN) in nats (0 = uniform,
    /// does not move the lane).
    public func laneDiagnostics(for word: String)
        -> (
            frequencyIS: UInt32?, frequencyEN: UInt32?, zIS: Double, zEN: Double,
            binKnown: Bool, evidence: Double
        )
    {
        let w = word.lowercased()
        return (
            frequencyIS: model.icelandic.frequency(of: w),
            frequencyEN: model.english.frequency(of: w),
            zIS: model.calibratedUnigramScore(of: w, language: .icelandic),
            zEN: model.calibratedUnigramScore(of: w, language: .english),
            binKnown: model.morphology?.isKnown(w) == true,
            evidence: model.laneEvidence(of: w)
        )
    }

    /// Touch representative pages of the mmap-ed artifacts (spread unigram,
    /// bigram and morphology lookups) so first keystrokes don't pay page
    /// faults. Call once after load, on the same queue that owns the engine.
    public func warmUp() {
        model.warmUp()
    }

    /// Reset the lane posterior to the neutral 50/50 prior — the full-decay
    /// case of the boundary relaxation (new text field, session reset).
    public func resetLanguagePosterior() {
        probabilityIcelandic = 0.5
    }

    // MARK: - Internals

    /// Trailing word of the committed context, lowercased and stripped of
    /// surrounding punctuation; nil if there is none.
    static func lastWord(in context: String) -> String? {
        guard
            let token = context
                .split(whereSeparator: { $0.isWhitespace || $0.isNewline })
                .last
        else { return nil }
        let stripped = token.trimmingCharacters(
            in: CharacterSet.punctuationCharacters.union(.symbols)
        )
        guard !stripped.isEmpty else { return nil }
        return stripped.lowercased()
    }
}

File: AndroidClient/lib/engine/src/main/kotlin/is/solberg/lyklabord/engine/TypeEngine.kt
package `is`.solberg.lyklabord.engine

import `is`.solberg.lyklabord.engine.config.EngineConfig
import `is`.solberg.lyklabord.engine.lexicon.Lexicon
import `is`.solberg.lyklabord.engine.lexicon.LexiconCalibrationProfile
import `is`.solberg.lyklabord.engine.morph.BinaryLemmatizer
import `is`.solberg.lyklabord.engine.morph.MorphologyProviding
import `is`.solberg.lyklabord.engine.morph.ParadigmBundle
import kotlin.math.exp
import kotlin.math.max
import kotlin.math.min

/**
 * Facade over the corrector and predictor with a running bilingual language
 * posterior. Calls are synchronous and must be confined to one owner queue.
 */
class TypeEngine(
    icelandic: Lexicon,
    english: Lexicon,
    morphologyProvider: MorphologyProviding? = null,
    val config: EngineConfig = EngineConfig(),
    icelandicCalibration: LexiconCalibrationProfile? = null,
    englishCalibration: LexiconCalibrationProfile? = null,
) {
    private val model = BlendedLanguageModel(
        icelandic = icelandic,
        english = english,
        morphology = morphologyProvider,
        config = config,
        icelandicCalibrationProfile = icelandicCalibration,
        englishCalibrationProfile = englishCalibration,
        personal = PersonalStore(),
    )
    private val corrector = Corrector(model = model, config = config)
    private val predictor = Predictor(model = model, config = config)

    /** Lane posterior P(Icelandic), initialized to the neutral prior. */
    var probabilityIcelandic: Double = 0.5
        private set

    /** Production morphology overload; BinaryLemmatizer already implements the seam. */
    constructor(
        icelandic: Lexicon,
        english: Lexicon,
        morphology: BinaryLemmatizer?,
        config: EngineConfig = EngineConfig(),
        icelandicCalibration: LexiconCalibrationProfile? = null,
        englishCalibration: LexiconCalibrationProfile? = null,
    ) : this(
        icelandic = icelandic,
        english = english,
        morphologyProvider = morphology,
        config = config,
        icelandicCalibration = icelandicCalibration,
        englishCalibration = englishCalibration,
    )

    data class CalibrationDiagnostics(
        val icelandicMean: Double,
        val icelandicSigma: Double,
        val icelandicWarmupWords: List<String>,
        val englishMean: Double,
        val englishSigma: Double,
        val englishWarmupWords: List<String>,
    )

    val calibrationDiagnostics: CalibrationDiagnostics
        get() = CalibrationDiagnostics(
            icelandicMean = model.icelandicCalibration.meanLogFrequency,
            icelandicSigma = model.icelandicCalibration.stdLogFrequency,
            icelandicWarmupWords = model.icelandicCalibration.sampleWords,
            englishMean = model.englishCalibration.meanLogFrequency,
            englishSigma = model.englishCalibration.stdLogFrequency,
            englishWarmupWords = model.englishCalibration.sampleWords,
        )

    fun setPersonalVocabulary(vocabulary: PersonalVocabulary?) {
        model.personal.setSnapshot(vocabulary)
        rebuildLemmaLift()
    }

    fun setPersonalTouch(snapshot: PersonalTouchSnapshot?) {
        model.touch.setSnapshot(snapshot)
    }

    val personalTouchSnapshot: PersonalTouchSnapshot?
        get() = model.touch.snapshot

    fun learnSessionWord(word: String) {
        model.personal.learnSession(word)
        rebuildLemmaLift()
    }

    fun clearSessionVocabulary() {
        model.personal.clearSession()
        rebuildLemmaLift()
    }

    fun forgetSessionWord(word: String) {
        model.personal.forgetSession(word)
        rebuildLemmaLift()
    }

    fun setInflection(inflection: InflectionModel?) {
        model.inflection.setModel(inflection)
        model.compounds.clearCache()
        rebuildLemmaLift()
    }

    private fun rebuildLemmaLift() {
        model.inflection.rebuildLift(
            words = model.personal.allValidKeys,
            morphology = model.morphology,
            liftNats = config.lemmaLiftBoost,
        )
    }

    data class GovernorDiagnostics(
        val mass: Double,
        val entropyRatio: Double,
        val cases: List<CaseProbability>,
    )

    data class CaseProbability(val name: String, val p: Double)

    fun governorDiagnostics(word: String): GovernorDiagnostics? {
        val governor = model.inflection.model?.governors?.governor(word.lowercase()) ?: return null
        val cases = ParadigmBundle.caseNames.mapIndexed { index, name ->
            CaseProbability(name, governor.caseProbabilities[index])
        }.filter { it.p > 0.0 }.sortedByDescending { it.p }
        return GovernorDiagnostics(governor.mass, governor.caseEntropyRatio, cases)
    }

    val personalSnapshotWords: List<String>
        get() = model.personal.snapshotWords

    val sessionLearnedWords: List<String>
        get() = model.personal.sessionWords

    fun isPersonalWord(word: String): Boolean = model.personal.isValidWord(word)

    fun isPersonalLearnedWord(word: String): Boolean =
        model.personal.isValidWord(word) &&
            !model.personal.isTombstoned(word) &&
            !model.isKnownAnywhere(word) &&
            model.compoundSplit(word) == null

    internal fun autocapArtifactLowercased(word: String): String? {
        if (word.length < 2) return null
        val first = word.firstOrNull() ?: return null
        if (!first.isUpperCase()) return null
        val lowered = word.lowercase()
        if (word != lowered.replaceRange(0, 1, lowered.first().uppercase())) return null
        if (!model.isCapArtifactBase(lowered)) return null
        return lowered
    }
    /** Direct correction entry point for embedders that do not need surface restoration. */
    fun correct(
        typed: String,
        previousWord: String? = null,
        limit: Int = 3,
        deliberateCharacters: List<Char> = emptyList(),
        taps: List<TapSample?> = emptyList(),
        capitalizedMidSentence: Boolean = false,
        trace: CorrectionTrace? = null,
    ): CorrectionResult = corrector.correct(
        typed = typed,
        previousWord = previousWord,
        pIcelandic = probabilityIcelandic,
        limit = limit,
        deliberateCharacters = deliberateCharacters,
        taps = taps,
        capitalizedMidSentence = capitalizedMidSentence,
        trace = trace,
    )

    /** Suggestion-bar entry point using the full facade behavior. */
    fun suggest(
        `context`: String,
        currentWord: String,
        limit: Int = 3,
        deliberateCharacters: List<Char> = emptyList(),
        taps: List<TapSample?> = emptyList(),
        trace: CorrectionTrace? = null,
    ): List<Suggestion> = suggestions(
        context = `context`,
        currentWord = currentWord,
        limit = limit,
        deliberateCharacters = deliberateCharacters,
        taps = taps,
        trace = trace,
    )

    /** Next-word prediction entry point. */
    fun predict(previousWord: String? = null, limit: Int = 3): List<Suggestion> =
        restorePersonalSurfaces(predictor.nextWords(previousWord, probabilityIcelandic, limit))

    /** Create a stateful editor session owned by this engine. */
    fun makeSession(): TypingSession = TypingSession(this)


    fun suggestions(
        `context`: String,
        currentWord: String,
        limit: Int = 3,
        deliberateCharacters: List<Char> = emptyList(),
        taps: List<TapSample?> = emptyList(),
        trace: CorrectionTrace? = null,
    ): List<Suggestion> {
        val previous = lastWord(`context`)
        val trimmed = currentWord.trim { it.isWhitespace() }
        if (trimmed.isEmpty()) {
            trace?.rule = "prediction (no word in progress)"
            return restorePersonalSurfaces(
                predictor.nextWords(
                    previousWord = previous,
                    pIcelandic = probabilityIcelandic,
                    limit = limit,
                ),
            )
        }
        val result = corrector.correct(
            typed = trimmed,
            previousWord = previous,
            pIcelandic = probabilityIcelandic,
            limit = limit,
            deliberateCharacters = deliberateCharacters,
            taps = if (taps.size == trimmed.length) taps else emptyList(),
            capitalizedMidSentence = trimmed.firstOrNull()?.isUpperCase() == true && isMidSentence(`context`),
            trace = trace,
        )
        var suggestions = restorePersonalSurfaces(result.suggestions)
        if (trimmed.firstOrNull()?.isUpperCase() == true) {
            suggestions = suggestions.map { suggestion ->
                val text = suggestion.text
                Suggestion(
                    text = if (text.isNotEmpty()) text.first().uppercase() + text.drop(1) else text,
                    isAutocorrect = suggestion.isAutocorrect,
                    confidence = suggestion.confidence,
                    isVerbatim = suggestion.isVerbatim,
                    isRestoration = suggestion.isRestoration,
                    isPersonalLearned = suggestion.isPersonalLearned,
                )
            }
        }
        return suggestions.map { suggestion ->
            if (suggestion.isAutocorrect && suggestion.text == trimmed) {
                suggestion.copy(isAutocorrect = false)
            } else suggestion
        }
    }

    fun dottedSpaceMiss(left: String, right: String, `context`: String): Suggestion? {
        val previous = lastWord(`context`)
        val suggestion = corrector.dotSplitSuggestion(
            left = left.lowercase(),
            right = right.lowercase(),
            previousWord = previous,
            pIcelandic = probabilityIcelandic,
        ) ?: return null
        if (left.firstOrNull()?.isUpperCase() == true) {
            val text = suggestion.text
            return suggestion.copy(
                text = if (text.isNotEmpty()) text.first().uppercase() + text.drop(1) else text,
            )
        }
        return suggestion
    }

    private fun restorePersonalSurfaces(suggestions: List<Suggestion>): List<Suggestion> = suggestions.map { suggestion ->
        val surface = model.personal.displaySurface(suggestion.text) ?: return@map suggestion
        val isAllCaps = surface.length > 1 && surface == surface.uppercase() && surface != surface.lowercase()
        if (isAllCaps) {
            val lowered = surface.lowercase()
            val baseAttested =
                model.icelandic.frequency(lowered) != null || model.english.frequency(lowered) != null
            if (baseAttested) return@map suggestion
        }
        val isLeadingCapOnly = !isAllCaps && surface == leadingCapital(suggestion.text)
        if (isLeadingCapOnly && model.isCapArtifactBase(suggestion.text)) return@map suggestion
        suggestion.copy(text = surface)
    }

    fun confirmWord(word: String) {
        val w = word.lowercase().trim { it.isWhitespace() }
        if (w.isEmpty()) return
        val switch = config.laneSwitchProbability
        val predicted = (1 - switch) * probabilityIcelandic + switch * (1 - probabilityIcelandic)
        val evidence = model.laneEvidence(w)
        val odds = (predicted / (1 - predicted)) * exp(evidence)
        val updated = odds / (1 + odds)
        probabilityIcelandic = min(max(updated, config.posteriorFloor), config.posteriorCeiling)
    }

    fun noteSentenceBoundary() {
        val decay = config.laneBoundaryDecay
        probabilityIcelandic = 0.5 + (probabilityIcelandic - 0.5) * (1 - decay)
    }

    data class LaneDiagnostics(
        val frequencyIS: UInt?,
        val frequencyEN: UInt?,
        val zIS: Double,
        val zEN: Double,
        val binKnown: Boolean,
        val binCases: List<String>,
        val binLemmas: List<String>,
        val evidence: Double,
    )

    fun laneDiagnostics(word: String): LaneDiagnostics {
        val w = word.lowercase()
        return LaneDiagnostics(
            frequencyIS = model.icelandic.frequency(w),
            frequencyEN = model.english.frequency(w),
            zIS = model.calibratedUnigramScore(w, Language.icelandic),
            zEN = model.calibratedUnigramScore(w, Language.english),
            binKnown = model.morphology?.isKnown(w) == true,
            binCases = model.morphology?.nounAdjectiveCases(w) ?: emptyList(),
            binLemmas = model.morphology?.lemmaCandidates(w) ?: emptyList(),
            evidence = model.laneEvidence(w),
        )
    }

    data class BigramDiagnostics(
        val bigramIS: UInt?,
        val bigramEN: UInt?,
        val previousIS: UInt?,
        val previousEN: UInt?,
        val zIS: Double,
        val zEN: Double,
    )

    fun bigramDiagnostics(previous: String, word: String): BigramDiagnostics {
        val p = previous.lowercase()
        val w = word.lowercase()
        return BigramDiagnostics(
            bigramIS = model.icelandic.bigramFrequency(p, w),
            bigramEN = model.english.bigramFrequency(p, w),
            previousIS = model.icelandic.frequency(p),
            previousEN = model.english.frequency(p),
            zIS = model.calibratedScore(w, p, Language.icelandic),
            zEN = model.calibratedScore(w, p, Language.english),
        )
    }

    data class CompoundDiagnostics(
        val typedValid: Boolean,
        val `protected`: Boolean,
        val parts: List<String>?,
        val deniedSplit: String?,
    )

    fun compoundDiagnostics(word: String): CompoundDiagnostics {
        val w = word.lowercase()
        return CompoundDiagnostics(
            typedValid = model.isValidTypedWord(w),
            `protected` = model.isProtectedTypedWord(w),
            parts = model.compoundSplit(w)?.let { it.modifiers + listOf(it.head) },
            deniedSplit = CompoundAnalyzer.neverCompounds[w],
        )
    }

    fun warmUp() {
        model.warmUp()
    }

    fun resetLanguagePosterior() {
        probabilityIcelandic = 0.5
    }

    companion object {
        fun isMidSentence(`context`: String): Boolean {
            val last = `context`.trim { it.isWhitespace() }.lastOrNull() ?: return false
            return last !in ".!?..."
        }

        fun lastWord(`context`: String): String? {
            val token = `context`.split(Regex("\\s+")).filter { it.isNotEmpty() }.lastOrNull() ?: return null
            val stripped = token.trim { it.isWhitespace() || it.category.isPunctuationOrSymbol() }
            return stripped.takeIf { it.isNotEmpty() }?.lowercase()
        }

        private fun leadingCapital(text: String): String =
            if (text.isEmpty()) text else text.first().uppercase() + text.drop(1)

        private fun CharCategory.isPunctuationOrSymbol(): Boolean =
            this == CharCategory.CONNECTOR_PUNCTUATION ||
                this == CharCategory.DASH_PUNCTUATION ||
                this == CharCategory.START_PUNCTUATION ||
                this == CharCategory.END_PUNCTUATION ||
                this == CharCategory.INITIAL_QUOTE_PUNCTUATION ||
                this == CharCategory.FINAL_QUOTE_PUNCTUATION ||
                this == CharCategory.OTHER_PUNCTUATION ||
                this == CharCategory.MATH_SYMBOL ||
                this == CharCategory.CURRENCY_SYMBOL ||
                this == CharCategory.MODIFIER_SYMBOL ||
                this == CharCategory.OTHER_SYMBOL
    }
}
